File: Example-ObjC/Simple/ViewController.m
//
//  ViewController.m
//  Simple
//
//  Created by Sedlak, Stefan on 7/21/16.
//  Copyright © 2016 Wirecard Technologies GmbH. All rights reserved.
//

#import "ViewController.h"

#import <PassKit/PassKit.h>
#import <PassKit/PKPaymentAuthorizationViewController.h>

#import <libextobjc/EXTScope.h>
#import <WDeCom/WDeCom.h>
#import <WDeComApplePay/WDeComApplePay.h>
#import <WDeComCard/WDeComCard.h>
#import <WDeComPayPal/WDeComPayPal.h>
#import <WDeComSEPA/WDeComSEPA.h>
#import <WDeComKlarna/WDeComKlarna.h>

#define AMOUNT [NSDecimalNumber decimalNumberWithMantissa:199 exponent:-2 isNegative:NO]

NSString *const PMTitleApplePay                 = @"ApplePay";
NSString *const PMTitleCard                     = @"Card";
NSString *const PMTitleCardManualBrandSelection = @"Card manual brand selection";
NSString *const PMTitlePayPal                   = @"PayPal";
NSString *const PMTitleSEPA                     = @"SEPA";
NSString *const PMTitleKlarna                   = @"Klarna";

@interface ViewController ()

@property (weak, nonatomic) IBOutlet UIButton *payBtn;

@end

@implementation ViewController

#pragma mark - Overriden

- (void)viewDidLoad {
    [super viewDidLoad];
    if (!self.client) {
        self.payBtn.enabled = NO;
    }
}

#pragma marks - Action

- (IBAction)onPay:(UIButton *)sender {
    UIAlertController *ac = [UIAlertController alertControllerWithTitle:@"Payment method"
                                                                message:nil
                                                         preferredStyle:UIAlertControllerStyleActionSheet];
    
    void (^handler)(UIAlertAction *action) = ^(UIAlertAction * _Nonnull action) {
        NSString *title = action.title;
        WDECPayment *payment = [self createPayment:title];
        if (!payment) {
            return;
        }
        sender.enabled = NO;
        @weakify(self, sender);
        [self.client makePayment:payment withCompletion:^(WDECPaymentResponse * _Nullable response, NSError * _Nullable error) {
            @strongify(self, sender);
            sender.enabled = YES;
            
            NSString *alertTitle = error ? @"Error" : @"Success";
            NSString *alertMessage = error ? error.localizedDescription : @"Item(s) has been purchased.";
            UIAlertController *ac = [UIAlertController alertControllerWithTitle:alertTitle
                                                                        message:alertMessage
                                                                 preferredStyle:UIAlertControllerStyleAlert];
            
            [ac addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
            [self presentViewController:ac animated:YES completion:nil];
        }];
    };
    
    BOOL canMakeApplePayPayment = [self canMakeApplePayPayment];
    if (canMakeApplePayPayment) {
        [ac addAction:[UIAlertAction actionWithTitle:PMTitleApplePay style:UIAlertActionStyleDefault handler:handler]];
    }
    [ac addAction:[UIAlertAction actionWithTitle:PMTitleCard                        style:UIAlertActionStyleDefault handler:handler]];
    [ac addAction:[UIAlertAction actionWithTitle:PMTitleCardManualBrandSelection    style:UIAlertActionStyleDefault handler:handler]];
    [ac addAction:[UIAlertAction actionWithTitle:PMTitlePayPal                      style:UIAlertActionStyleDefault handler:handler]];
    [ac addAction:[UIAlertAction actionWithTitle:PMTitleSEPA                        style:UIAlertActionStyleDefault handler:handler]];
    [ac addAction:[UIAlertAction actionWithTitle:PMTitleKlarna                      style:UIAlertActionStyleDefault handler:handler]];
    [ac addAction:[UIAlertAction actionWithTitle:@"Cancel"                          style:UIAlertActionStyleCancel handler:handler]];
    
    [self presentViewController:ac animated:YES completion:nil];
}

#pragma mark - Private

- (BOOL)canMakeApplePayPayment
{
    NSArray<NSString *> *networks = @[PKPaymentNetworkMasterCard, PKPaymentNetworkVisa];
    BOOL result = [PKPaymentAuthorizationViewController canMakePaymentsUsingNetworks:networks];
    return result;
}

- (WDECPayment *)createPayment:(nonnull NSString *)title
{
    WDECPayment *result = nil;
    if ([title isEqualToString:PMTitleApplePay]) {
        result = [self createPaymentApplePay];
    } else if ([title isEqualToString:PMTitleCard]) {
        result = [self createPaymentCard];
        [[WDECCardLayout appearance] setManualCardBrandSelectionRequired:NO];
    } else if ([title isEqualToString:PMTitleCardManualBrandSelection]) {
        result = [self createPaymentCard];
        [[WDECCardLayout appearance] setManualCardBrandSelectionRequired:YES];
    } else if ([title isEqualToString:PMTitlePayPal]) {
        result = [self createPaymentPayPal];
    } else if ([title isEqualToString:PMTitleSEPA]) {
        result = [self createPaymentSEPA];
    } else if ([title isEqualToString:PMTitleKlarna]) {
        result = [self createPaymentKlarna];
    }
    return result;
}

- (WDECPayment *)createPaymentApplePay
{
    static NSString *const APPLE_MERCHANT_ID = @"merchant.com.wirecard.paymentsdk.example.Simple";
    WDECApplePayManagedPayment *payment = [WDECApplePayManagedPayment new];
    payment.appleMerchantID = APPLE_MERCHANT_ID;
    payment.appleMerchantCountry = WDECCountryUS;
    payment.amount = AMOUNT;
    payment.currency = @"USD";
    payment.transactionType = WDECTransactionTypePurchase;
    
    static NSString *const WD_MERCHANT_ACCOUNT_ID = @"";
    static NSString *const WD_MERCHANT_SECRET_KEY = @"";
    /**
    @Warning: We ask you to never share your Secret Key with anyone, or store it inside your application or phone. This is crucial to ensure the security of your transactions.
    You will be generating the signature on your own server’s backend, as it is the only place where you will store your Secret Key.
    */
    [self merchant:WD_MERCHANT_ACCOUNT_ID signPayment:payment byMerchantSecretKey:WD_MERCHANT_SECRET_KEY];
    return payment;
}

- (WDECPayment *)createPaymentCard
{
    WDECCardPayment *payment = [WDECCardPayment new];
    payment.amount = AMOUNT;
    payment.currency = @"USD";
    payment.transactionType = WDECTransactionTypePurchase;
    /**
    @Warning: We ask you to never share your Secret Key with anyone, or store it inside your application or phone. This is crucial to ensure the security of your transactions.
    You will be generating the signature on your own server’s backend, as it is the only place where you will store your Secret Key.
    */
    static NSString *const WD_MERCHANT_ACCOUNT_ID = @"33f6d473-3036-4ca5-acb5-8c64dac862d1";
    static NSString *const WD_MERCHANT_SECRET_KEY = @"9e0130f6-2e1e-4185-b0d5-dc69079c75cc";
    [self merchant:WD_MERCHANT_ACCOUNT_ID signPayment:payment byMerchantSecretKey:WD_MERCHANT_SECRET_KEY];
    return payment;
}

- (WDECPayment *)createPaymentPayPal
{
    WDECOrderItem *orderItem = [WDECOrderItem new];
    orderItem.name = @"The Watch";
    orderItem.amount = AMOUNT;
    orderItem.amountCurrency = WDECCurrencyEUR;
    
    WDECOrder *order = [WDECOrder new];
    order.descriptor = @"DEMO DESCRIPTOR";
    order.number = [[NSUUID UUID] UUIDString];
    order.detail = @"DEMO ORDER DETAIL";
    order.items = @[orderItem];
    
    WDECPayPalPayment *payment = [WDECPayPalPayment new];
    payment.amount = AMOUNT;
    payment.currency = @"EUR";
    payment.transactionType = WDECTransactionTypeDebit;
    payment.order = order;
    /**
    @Warning: We ask you to never share your Secret Key with anyone, or store it inside your application or phone. This is crucial to ensure the security of your transactions.
    You will be generating the signature on your own server’s backend, as it is the only place where you will store your Secret Key.
    */
    static NSString *const WD_MERCHANT_ACCOUNT_ID = @"9abf05c1-c266-46ae-8eac-7f87ca97af28";
    static NSString *const WD_MERCHANT_SECRET_KEY = @"5fca2a83-89ca-4f9e-8cf7-4ca74a02773f";
    [self merchant:WD_MERCHANT_ACCOUNT_ID signPayment:payment byMerchantSecretKey:WD_MERCHANT_SECRET_KEY];
    payment.accountHolder = [self accountHolder];
    
    return payment;
}

- (WDECPayment *)createPaymentSEPA
{
    WDECSEPAPayment *payment = [WDECSEPAPayment new];
    payment.creditorID = @"DE98ZZZ09999999999";
    payment.mandateID = @"12345678";
    payment.amount = AMOUNT;
    payment.currency = @"EUR";
    payment.transactionType = WDECTransactionTypePendingDebit;
    /**
    @Warning: We ask you to never share your Secret Key with anyone, or store it inside your application or phone. This is crucial to ensure the security of your transactions.
    You will be generating the signature on your own server’s backend, as it is the only place where you will store your Secret Key.
    */
    static NSString *const WD_MERCHANT_ACCOUNT_ID = @"4c901196-eff7-411e-82a3-5ef6b6860d64";
    static NSString *const WD_MERCHANT_SECRET_KEY = @"ecdf5990-0372-47cd-a55d-037dccfe9d25";
    [self merchant:WD_MERCHANT_ACCOUNT_ID signPayment:payment byMerchantSecretKey:WD_MERCHANT_SECRET_KEY];
    return payment;
}

- (WDECPayment *)createPaymentKlarna
{
    WDECLocalize localize = [WDECLocalize appearance];
    BOOL locale_de = localize.locale == WDECLocalize_de;

    WDECCountry country = locale_de ? WDECCountryDE : WDECCountryGB;

    WDECKlarnaPayment *payment = [WDECKlarnaPayment new];
    payment.amount = AMOUNT;
    payment.currency = locale_de ? WDECCurrencyGetISOCode(WDECCurrencyEUR) : WDECCurrencyGetISOCode(WDECCurrencyGBP);
    payment.transactionType = WDECTransactionTypeAuthorization;
    payment.category = WDECKlarnaCategoryDebit;
    payment.locale = locale_de ? WDECLocalize_de : WDECLocale_en;
    /**
    @Warning: We ask you to never share your Secret Key with anyone, or store it inside your application or phone. This is crucial to ensure the security of your transactions.
    You will be generating the signature on your own server’s backend, as it is the only place where you will store your Secret Key.
    */
    static NSString *const WD_MERCHANT_ACCOUNT_ID = @"f570c123-62f1-4a0d-8688-d999a05d50d4";
    static NSString *const WD_MERCHANT_SECRET_KEY = @"0fb50d2c-8ab5-4d53-ac69-b707b1319148";
    [self merchant:WD_MERCHANT_ACCOUNT_ID signPayment:payment byMerchantSecretKey:WD_MERCHANT_SECRET_KEY];
    
    payment.category = WDECKlarnaCategoryPayLater;
    payment.appScheme = @"iPay.Klarna";

    WDECCustomerData *accountHolder = [WDECCustomerData new];
    
    WDECAddress* address = [WDECAddress new];
    address.country = country;
    accountHolder.address = address;

    payment.accountHolder = accountHolder;

    WDECOrderItem *order1 = [WDECOrderItem new];
    order1.quantity = [NSDecimalNumber decimalNumberWithString:@"1"];
    order1.name =  @"Item 1";
    order1.articleNumber =  @"1";
    order1.amount =  [NSDecimalNumber decimalNumberWithString:@"20"];
    order1.amountCurrency = locale_de ? WDECCurrencyDE : WDECCurrencyGBP;
    order1.type = WDECOrderItemTypePhysical;

    WDECOrderItem *order2 = [WDECOrderItem new];
    order2.quantity = [NSDecimalNumber decimalNumberWithString:@"1"];
    order2.name =  @"Item 2";
    order2.articleNumber =  @"2";
    order2.amount =  [NSDecimalNumber decimalNumberWithString:@"5.0"];
    order2.amountCurrency = locale_de ? WDECCurrencyDE : WDECCurrencyGBP;
    order2.type = WDECOrderItemTypeDiscount;

    payment.order = [WDECOrder new];
    payment.order.items = @[order1, order2];

    return payment;
}

- (WDECCustomerData *)accountHolder {
    WDECCustomerData *accountHolder = nil;
    accountHolder = [WDECCustomerData new];
    accountHolder.lastName = @"Doe";
    
    return accountHolder;
}

@end

File: Example-Swift/Simple/ViewController.swift
//
//  ViewController.swift
//  Simple
//
//  Created by Vrana, Jozef on 2/21/17.
//  Copyright © 2017 Vrana, Jozef. All rights reserved.
//

import UIKit
import PassKit
import WDeComApplePay
import WDeCom
import WDeComCard
import WDeComPayPal
import WDeComSEPA
import WDeComKlarna

let AMOUNT = NSDecimalNumber.init(mantissa: 199, exponent: -2, isNegative: false)

let PMTitleApplePay = "ApplePay"
let PMTitleCard     = "Card"
let PMTitleCardManualBrandSelection = "Card manual brand selection"
let PMTitlePayPal   = "PayPal"
let PMTitleSEPA     = "SEPA"
let PMTitleKlarna   = "Klarna"

class ViewController: PaymemtVC, UIActionSheetDelegate, UIPopoverPresentationControllerDelegate {
    
    @IBOutlet weak var payButton: UIButton!
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    @IBAction func onPay(_ sender: UIButton) {
    
        let actionSheetController = UIAlertController(title: "Payment", message: nil, preferredStyle: .actionSheet)
        
        func handler(actionTarget: UIAlertAction){
            
            
            if actionTarget.style == .cancel {
                self.dismiss(animated: true, completion: nil)
                return
            }
            
            let paymentTitle = actionTarget.title
            let payment = self.createPayment(title: paymentTitle!)
            
            if self.client == nil {
                return
            }
           
            self.client!.make(payment, withCompletion:{(response: WDECPaymentResponse?,error: Error?) in
                let alertTitle = error != nil ? "Error" : "Success"
                let alertMessage = error != nil ? error!.localizedDescription : "Item(s) has been purchased."
                let ac = UIAlertController(title: alertTitle, message: alertMessage, preferredStyle: .actionSheet)
                ac.popoverPresentationController?.sourceView = self.view
                ac.popoverPresentationController?.sourceRect = CGRect(x: self.view.bounds.width / 2.0,y: self.view.bounds.height,width: 1.0,height: 1.0)

                ac.addAction(UIAlertAction(title:"OK", style:.default, handler:nil));
                self.present(ac, animated:true, completion:nil)
            })
        }
        let canMakeApplePayPayment = self.canMakeApplePayPayment()
        if canMakeApplePayPayment {
            actionSheetController.addAction(UIAlertAction(title: PMTitleApplePay, style: .default, handler: handler))
        }
        actionSheetController.addAction(UIAlertAction(title: PMTitleCard, style: .default, handler: handler))
        actionSheetController.addAction(UIAlertAction(title: PMTitleCardManualBrandSelection, style: .default, handler: handler))
        actionSheetController.addAction(UIAlertAction(title: PMTitlePayPal, style: .default, handler: handler))
        actionSheetController.addAction(UIAlertAction(title: PMTitleSEPA, style: .default, handler: handler))
        actionSheetController.addAction(UIAlertAction(title: PMTitleKlarna, style: .default, handler: handler))
        actionSheetController.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: handler))
        actionSheetController.popoverPresentationController?.sourceView = self.view
        actionSheetController.popoverPresentationController?.sourceRect = CGRect(x: self.view.bounds.width / 2.0,y: self.view.bounds.height,width: 1.0,height: 1.0)
        
        self.present(actionSheetController, animated: true, completion: nil)
    }
    
    func prepareForPopoverPresentation(_ popoverPresentationController: UIPopoverPresentationController) {
        
    }
    
    func canMakeApplePayPayment() -> Bool {
        let networks = [PKPaymentNetwork.masterCard, PKPaymentNetwork.visa]
        let result = PKPaymentAuthorizationViewController.canMakePayments(usingNetworks: networks)
        return result;
    }
    
    func createPayment(title:String) -> WDECPayment {
        var result : WDECPayment
        switch title {
        case PMTitleApplePay : result = self.createPaymentApplePay()
        case PMTitleCard : result = self.createPaymentCard()
        let cardLayout =  WDECCardLayout.appearance()
        cardLayout.manualCardBrandSelectionRequired = false
        case PMTitleCardManualBrandSelection :
            result = self.createPaymentCard()
            let cardLayout =  WDECCardLayout.appearance()
            cardLayout.manualCardBrandSelectionRequired = true
        case PMTitlePayPal : result = self.createPaymentPayPal()
        case PMTitleSEPA : result = self.createPaymentSEPA()
        case PMTitleKlarna: result = self.createPaymentKlarna()
        default : result = WDECPayment()
        }
        return result
    }
    
    func createPaymentApplePay() -> WDECPayment {
        
        let APPLE_MERCHANT_ID = "merchant.com.wirecard.paymentsdk.example.Simple"
        let payment = WDECApplePayManagedPayment()
        payment.appleMerchantID = APPLE_MERCHANT_ID
        payment.appleMerchantCountry = .US
        payment.amount = AMOUNT
        payment.currency = WDECCurrencyGetISOCode(WDECCurrency.USD)
        payment.transactionType = .purchase
        /**
        @Warning: We ask you to never share your Secret Key with anyone, or store it inside your application or phone. This is crucial to ensure the security of your transactions.
        You will be generating the signature on your own server’s backend, as it is the only place where you will store your Secret Key.
        */
        let WD_MERCHANT_ACCOUNT_ID = "";
        let WD_MERCHANT_SECRET_KEY = "";
        self.merchantSignedPaymentByMerchantSecretKey(merchantAccountID: WD_MERCHANT_ACCOUNT_ID, payment: payment, merchantSecretKey: WD_MERCHANT_SECRET_KEY)
        
        return payment
    }
    
    func createPaymentCard() -> WDECPayment {
        let payment = WDECCardPayment()
        payment.amount = AMOUNT
        payment.currency = "USD"
        payment.transactionType = .purchase
        /**
        @Warning: We ask you to never share your Secret Key with anyone, or store it inside your application or phone. This is crucial to ensure the security of your transactions.
        You will be generating the signature on your own server’s backend, as it is the only place where you will store your Secret Key.
        */
        let WD_MERCHANT_ACCOUNT_ID = "33f6d473-3036-4ca5-acb5-8c64dac862d1"
        let WD_MERCHANT_SECRET_KEY = "9e0130f6-2e1e-4185-b0d5-dc69079c75cc"
        self.merchantSignedPaymentByMerchantSecretKey(merchantAccountID: WD_MERCHANT_ACCOUNT_ID, payment: payment, merchantSecretKey: WD_MERCHANT_SECRET_KEY)
        
        return payment
    }
    
    func createPaymentPayPal() -> WDECPayment {
        let orderItem = WDECOrderItem()
        orderItem.name = "The Watch"
        orderItem.amount = AMOUNT;
        orderItem.amountCurrency = .EUR;
        
        let order = WDECOrder()
        order.descriptor = "DEMO DESCRIPTOR"
        order.number = NSUUID.init().uuidString
        order.detail = "DEMO ORDER DETAIL"
        order.items = [orderItem]
        
        let payment = WDECPayPalPayment()
        payment.amount = AMOUNT
        payment.currency = "EUR"
        payment.transactionType = .debit
        payment.order = order;
        /**
        @Warning: We ask you to never share your Secret Key with anyone, or store it inside your application or phone. This is crucial to ensure the security of your transactions.
        You will be generating the signature on your own server’s backend, as it is the only place where you will store your Secret Key.
        */
        let WD_MERCHANT_ACCOUNT_ID = "9abf05c1-c266-46ae-8eac-7f87ca97af28"
        let WD_MERCHANT_SECRET_KEY = "5fca2a83-89ca-4f9e-8cf7-4ca74a02773f"
        self.merchantSignedPaymentByMerchantSecretKey(merchantAccountID: WD_MERCHANT_ACCOUNT_ID, payment: payment, merchantSecretKey: WD_MERCHANT_SECRET_KEY)
        
        return payment
    }
    
    func createPaymentSEPA() -> WDECPayment {
        let payment = WDECSEPAPayment()
        payment.creditorID = "DE98ZZZ09999999999"
        payment.mandateID = "12345678"
        payment.amount = AMOUNT
        payment.currency = "EUR"
        payment.transactionType = .pendingDebit
        /**
        @Warning: We ask you to never share your Secret Key with anyone, or store it inside your application or phone. This is crucial to ensure the security of your transactions.
        You will be generating the signature on your own server’s backend, as it is the only place where you will store your Secret Key.
        */
        let WD_MERCHANT_ACCOUNT_ID = "4c901196-eff7-411e-82a3-5ef6b6860d64"
        let WD_MERCHANT_SECRET_KEY = "ecdf5990-0372-47cd-a55d-037dccfe9d25"
        
        self.merchantSignedPaymentByMerchantSecretKey(merchantAccountID: WD_MERCHANT_ACCOUNT_ID, payment: payment, merchantSecretKey: WD_MERCHANT_SECRET_KEY)
        
        return payment
    }
    
    func createPaymentKlarna() -> WDECPayment {
        
        let localize = WDECLocalize.appearance()
        let locale_de = localize.locale == ._de;
        
        let payment = WDECKlarnaPayment()
        payment.amount = AMOUNT
        payment.currency = locale_de ? WDECCurrencyGetISOCode(WDECCurrency.EUR) : WDECCurrencyGetISOCode(WDECCurrency.GBP)
        payment.transactionType = WDECTransactionType.authorization
        payment.category = WDECKlarnaCategory.payLater;
        
        /**
        @Warning: We ask you to never share your Secret Key with anyone, or store it inside your application or phone. This is crucial to ensure the security of your transactions.
        You will be generating the signature on your own server’s backend, as it is the only place where you will store your Secret Key.
        */
        let WD_MERCHANT_ACCOUNT_ID = "f570c123-62f1-4a0d-8688-d999a05d50d4"
        let WD_MERCHANT_SECRET_KEY = "0fb50d2c-8ab5-4d53-ac69-b707b1319148"

        self.merchantSignedPaymentByMerchantSecretKey(merchantAccountID: WD_MERCHANT_ACCOUNT_ID, payment: payment, merchantSecretKey: WD_MERCHANT_SECRET_KEY)
                        
        let country = localize.locale == ._de ? WDECCountry.DE : WDECCountry.GB

        payment.transactionType = WDECTransactionType.authorization
        payment.locale = locale_de ? WDECLocale._de : WDECLocale._en;

        payment.category = WDECKlarnaCategory.payLater;
        payment.appScheme = "iPay.Klarna";

        let accountHolder = WDECCustomerData();
        let address = WDECAddress()
        address.country = country;
        accountHolder.address = address;

        payment.accountHolder = accountHolder;

        var cartOrder = [WDECOrderItem]()
        
        var orderItem = WDECOrderItem()
        orderItem.articleNumber = "1"
        orderItem.name = "Item 1"
        orderItem.amount = NSDecimalNumber(integerLiteral: 100)
        orderItem.quantity = NSDecimalNumber(integerLiteral: 1)
        orderItem.amountCurrency = locale_de ? WDECCurrency.EUR : WDECCurrency.GBP
        orderItem.type = WDECOrderItemType.physical

        cartOrder.append(orderItem)
        
        orderItem = WDECOrderItem()
        orderItem.articleNumber = "2"
        orderItem.name = "Item 2"
        orderItem.amount = NSDecimalNumber(integerLiteral: 99)
        orderItem.quantity = NSDecimalNumber(integerLiteral: 1)
        orderItem.amountCurrency = locale_de ? WDECCurrency.EUR : WDECCurrency.GBP
        orderItem.type = WDECOrderItemType.physical

        cartOrder.append(orderItem)
        
        let order = WDECOrder()
        order.descriptor = "DEMO DESCRIPTOR"
        order.number = NSUUID.init().uuidString
        order.detail = "DEMO ORDER DETAIL"
        order.items = cartOrder
        
        payment.order = order
        
        return payment
    }
}


